README.md
# Against996 README

Against 996 and refuse to work overtime

## Features

- The status bar shows commuting hours
- Customize the configuration commuting information

## Requirements

-

## Extension Settings

This extension contributes the following settings:

- `Against996.ignoreHolidays`: Whether holiday times are ignored
- `Against996.startTime`: The start time of work
- `Against996.endTime`: The end time of work
- `Against996.countDownCloseWindow`: The countdown to the end of the day, the countdown to the window before closing (in seconds)
- `Against996.startText`: The work prompt copy
- `Against996.startConfirmText`: The work confirmation button copy
- `Against996.endText`: The off-the-job prompt copy

## Known Issues

-

## Release Notes

Users appreciate release notes as you update your extension.

## v0.0.4 (2022-09-11)

### New feature:

- 增加是否忽略节假日([`83fe0ae`](https://github.com/Binbiubiubiu/vscode-against-996/commit/83fe0aece5c870b85157d045877865b1eb7d9ecb)) (by Binbiubiubiu)

## v0.0.3 (2022-09-10)

### Bugs fixed:

- 修复问题([`88b8030`](https://github.com/Binbiubiubiu/vscode-against-996/commit/88b803005bf5b2cd914ca455c9ab2509a4a08fa1)) (by Binbiubiubiu)

## v0.0.2 (2022-09-10)

### Bugs fixed:

- 修复打开应用校验([`54d0067`](https://github.com/Binbiubiubiu/vscode-against-996/commit/54d006745a4312deddfb0f1fe75c7f546a1e19b3)) (by Binbiubiubiu)

## v0.0.1 (2022-09-10)

### New feature:

- 添加 icon([`e72daac`](https://github.com/Binbiubiubiu/vscode-against-996/commit/e72daac54173f96627aea18104f2bea949e99495)) (by Binbiubiubiu)
- init project([`4d1b1ce`](https://github.com/Binbiubiubiu/vscode-against-996/commit/4d1b1ce4603d40fd35e907754cc42aa06842a18c)) (by kujiu)

### Bugs fixed:

- 修改配置([`5ac55cc`](https://github.com/Binbiubiubiu/vscode-against-996/commit/5ac55ccb4e16c1eaac981cf8176a99b635bf9011)) (by Binbiubiubiu)


CHANGELOG.md
# Change Log

All notable changes to the "Against996" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## v0.0.4 (2022-09-11)

### New feature:

- 增加是否忽略节假日([`83fe0ae`](https://github.com/Binbiubiubiu/vscode-against-996/commit/83fe0aece5c870b85157d045877865b1eb7d9ecb)) (by Binbiubiubiu)

## v0.0.3 (2022-09-10)

### Bugs fixed:

- 修复问题([`88b8030`](https://github.com/Binbiubiubiu/vscode-against-996/commit/88b803005bf5b2cd914ca455c9ab2509a4a08fa1)) (by Binbiubiubiu)

## v0.0.2 (2022-09-10)

### Bugs fixed:

- 修复打开应用校验([`54d0067`](https://github.com/Binbiubiubiu/vscode-against-996/commit/54d006745a4312deddfb0f1fe75c7f546a1e19b3)) (by Binbiubiubiu)

## v0.0.1 (2022-09-10)

### New feature:

- 添加 icon([`e72daac`](https://github.com/Binbiubiubiu/vscode-against-996/commit/e72daac54173f96627aea18104f2bea949e99495)) (by Binbiubiubiu)
- init project([`4d1b1ce`](https://github.com/Binbiubiubiu/vscode-against-996/commit/4d1b1ce4603d40fd35e907754cc42aa06842a18c)) (by kujiu)

### Bugs fixed:

- 修改配置([`5ac55cc`](https://github.com/Binbiubiubiu/vscode-against-996/commit/5ac55ccb4e16c1eaac981cf8176a99b635bf9011)) (by Binbiubiubiu)


src/api.ts
import * as dayjs from "dayjs";
import request from "./request";

const isWorkday = (d: string) => ["一", "二", "三", "四", "五"].includes(d);

export async function isHoliday(query: string) {
  let data: any;
  const now = new Date();
  const url = `https://opendata.baidu.com/api.php?tn=wisetpl&format=json&resource_id=39043&query=${encodeURIComponent(
    query
  )}&t=${now.getTime()}&cb=cb`;

  try {
    const res = await request.get(url);
    data = res.data;
  } catch (e) {
    return Promise.reject(e);
  }

  data = data.slice(7, data.length - 2);
  let dayInfo: any = {};
  try {
    data = JSON.parse(data);

    const { almanac = {} } = data.data[0];
    const d = now.getDate();
    dayInfo = almanac[d];
  } catch (e) {
    return Promise.reject("接口返回数据格式化异常");
  }

  if (dayInfo.status === 2) {
    // 法定补假日
    return false;
  } else if (dayInfo.status === 1) {
    // 法定节假日
    return true;
  }
  if (isWorkday(dayInfo.cnDay)) {
    return false;
  }
  return true;
}
export async function isTodayHoliday() {
  const now = dayjs();
  return isHoliday(`${now.year()}年${+now.month() + 1}月`);
}


src/monitor.ts
import * as vscode from "vscode";
import * as dayjs from "dayjs";
import * as schedule from "node-schedule";
import CONFIG from "./config";
import { isTodayHoliday } from "./api";

let isHoliday = false;
export async function initConfig() {
  if (CONFIG.ignoreHolidays) {
    isHoliday = false;
  } else {
    isHoliday = await isTodayHoliday();
  }
  console.log("initConfig", isHoliday);
  checkIsWorktime();
}
initConfig();

// 没到上班时间
async function isBeforeWorktime(now: dayjs.Dayjs) {
  const { hour, minute, second } = CONFIG.startTime;
  const startTime = dayjs().hour(hour).minute(minute).second(second);

  return now.isBefore(startTime);
}

async function beforeWorktimeAction(flag: boolean = false) {
  if (isHoliday || !flag) {
    return;
  }
  vscode.window
    .showWarningMessage(CONFIG.startText, CONFIG.startConfirmText)
    .then((rt) => rt && closeWindow());
}

// 已经下班了
async function isAfterWorktime(now: dayjs.Dayjs) {
  const { hour, minute, second } = CONFIG.endTime;
  const endTime = dayjs().hour(hour).minute(minute).second(second);

  return now.isAfter(endTime);
}

async function afterWorktimeAction(flag: boolean = false) {
  if (isHoliday || !flag) {
    return;
  }

  let time = +CONFIG.countDownCloseWindow;
  const endText = CONFIG.endText;
  const timeId = setInterval(() => {
    if (time === 0) {
      clearInterval(timeId);
      closeWindow();
      return;
    }
    vscode.window.showWarningMessage(
      endText.replace(/\{\{time\}\}/g, "" + time--)
    );
  }, 1000);
}

export async function checkIsWorktime() {
  const now = dayjs();
  isBeforeWorktime(now).then(beforeWorktimeAction);
  isAfterWorktime(now).then(afterWorktimeAction);
}
let startWorkJob: schedule.Job;
let endWorkJob: schedule.Job;
let isHolidayJob: schedule.Job;

export async function listen() {
  const st = CONFIG.startTime;
  startWorkJob = schedule.scheduleJob(
    `${st.second} ${st.minute} ${st.hour} * * ?`,
    () => beforeWorktimeAction(true)
  );
  const et = CONFIG.endTime;
  endWorkJob = schedule.scheduleJob(
    `${et.second} ${et.minute} ${et.hour} * * ?`,
    () => afterWorktimeAction(true)
  );
  isHolidayJob = schedule.scheduleJob("0 0 0 * * ?", initConfig);
}

export async function update() {
  if (startWorkJob) {
    const st = CONFIG.startTime;
    startWorkJob.reschedule(`${st.second} ${st.minute} ${st.hour} * * ?`);
  }

  if (endWorkJob) {
    const et = CONFIG.endTime;
    endWorkJob.reschedule(`${et.second} ${et.minute} ${et.hour} * * ?`);
  }
}

export async function close() {
  schedule.gracefulShutdown();
}

function closeWindow() {
  vscode.commands.executeCommand("workbench.action.closeWindow");
}


src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from "vscode";
import * as monitor from "./monitor";
import CONFIG, { handleConfigChangeListener } from "./config";

export let myStatusBarItem: vscode.StatusBarItem;

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate({ subscriptions }: vscode.ExtensionContext) {
  // Use the console to output diagnostic information (console.log) and errors (console.error)
  // This line of code will only be executed once when your extension is activated
  console.log('Congratulations, your extension "Against996" is now active!');

  const myCommandId = "Against996.against996";
  // The command has been defined in the package.json file
  // Now provide the implementation of the command with registerCommand
  // The commandId parameter must match the command field in package.json
  let disposable = vscode.commands.registerCommand(
    myCommandId,
    monitor.checkIsWorktime
  );

  subscriptions.push(disposable);
  // create a new status bar item that we can now manage
  myStatusBarItem = vscode.window.createStatusBarItem(
    vscode.StatusBarAlignment.Left,
    1
  );
  myStatusBarItem.command = myCommandId;
  myStatusBarItem.text = `Against 996(${CONFIG.startTime.value} - ${CONFIG.endTime.value})`;
  myStatusBarItem.show();
  subscriptions.push(myStatusBarItem);

  // update status bar item once at start
  monitor.listen();
  subscriptions.push(handleConfigChangeListener);
  monitor.checkIsWorktime();
}

// this method is called when your extension is deactivated
export function deactivate() {
  monitor.close();
}


src/config.ts
import * as vscode from "vscode";
import { myStatusBarItem } from "./extension";
import * as monitor from "./monitor";

let pluginConfig: vscode.WorkspaceConfiguration;

export function refreshConfig() {
  pluginConfig = vscode.workspace.getConfiguration("Against996");
}

refreshConfig();

export function getTimeConfig(key: string) {
  const config = pluginConfig.get<string>(key);
  if (!config) {
    return config as undefined;
  }
  const [hour, minute, second] = config.split(":");
  return {
    value: config,
    hour: +hour,
    minute: +minute,
    second: +second,
  };
}

const CONFIG = new Proxy(
  {
    ignoreHolidays: false,
    startTime: {
      value: "09:00:00",
      hour: 9,
      minute: 0,
      second: 0,
    },
    endTime: {
      value: "18:00:00",
      hour: 18,
      minute: 0,
      second: 0,
    },
    countDownCloseWindow: "5",
    startText: "宝贝,还没到上班时间,着急啥!",
    startConfirmText: "好的,不急",
    endText: "{{time}}s后强制关闭! 宝贝,还不赶紧下班!",
  },
  {
    get(target, key: string, receiver) {
      let v = undefined;
      const isTime = key.toString().toLowerCase().includes("time");
      if (isTime) {
        v = getTimeConfig(key);
      } else {
        v = pluginConfig.get(key);
      }
      return v ?? Reflect.get(target, key, receiver);
    },
  }
);

export default CONFIG;

export const handleConfigChangeListener =
  vscode.workspace.onDidChangeConfiguration((e) => {
    const isConfigChange = (...args: string[]) =>
      args.some((k) => e.affectsConfiguration(k));

    if (!isConfigChange("Against996")) {
      return;
    }
    refreshConfig();

    if (isConfigChange("Against996.startTime", "Against996.endTime")) {
      monitor.update();
      myStatusBarItem.text = `Against 996(${CONFIG.startTime.value} - ${CONFIG.endTime.value})`;
    }

    if (isConfigChange("Against996.ignoreHolidays")) {
      monitor.initConfig();
    }
  });
